src/operator/nn/mkldnn/mkldnn_softmax.cc

```diff
@@ -23,189 +23,211 @@
+  key.AddSign(axis);
   key.AddSign(is_train);
-  key.AddSign(data);
-  key.AddSign(output);
+  key.AddSign(temperature);
+  key.AddSign(tensors.data);
+  key.AddSign(tensors.out);
 
   auto it = fwds.find(key);
   if (it == fwds.end()) {
-    MKLDNNSoftmaxFwd fwd(is_train, real_axis, *(data.GetMKLDNNData()));
+    MKLDNNSoftmaxFwd fwd(param, tensors, is_train);
     it = AddToCache(&fwds, key, fwd);
   }
   return it->second;
 }
 
-void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
-                          const OpContext& ctx,
-                          const NDArray& in_data,
-                          const OpReqType& req,
-                          const NDArray& out_data) {
-  if (req == kNullOp)
-    return;
-  // same as the FCompute path, softmax only supports kWriteTo and kWriteInplace for now.
-  CHECK_NE(req, kAddTo);
+softmax_fwd_pd_t MKLDNNSoftmaxFwd::GetSoftmaxFwdPd(const mkldnn::memory& input_mem,
+                                                   const int axis,
+                                                   const bool is_train) {
+  const auto data_md    = input_mem.get_desc();
+  const auto cpu_engine = CpuEngine::Get()->get_engine();
+  const auto prop =
+      is_train ? mkldnn::prop_kind::forward_training : mkldnn::prop_kind::forward_scoring;
+  const auto desc = mkldnn::softmax_forward::desc(prop, data_md, axis);
+  return softmax_fwd_pd_t(desc, cpu_engine);
+}
 
-  const SoftmaxParam& param = nnvm::get<SoftmaxParam>(attrs.parsed);
-  int axis                  = CheckAxis(param.axis, in_data.shape().ndim());
-  auto fwd                  = GetSoftmaxFwd(param, axis, ctx.is_train, in_data, out_data);
+linear_pd_t MKLDNNSoftmaxFwd::GetTemperaturePd(const mkldnn::memory& input_mem,
+                                               const float temperature) {
+  const auto data_md    = input_mem.get_desc();
+  const auto cpu_engine = CpuEngine::Get()->get_engine();
+  const auto desc       = mkldnn::eltwise_forward::desc(mkldnn::prop_kind::forward_scoring,
+                                                  mkldnn::algorithm::eltwise_linear,
+                                                  data_md,
+                                                  1.0f / temperature,
+                                                  0.0f);
+  return linear_pd_t(desc, cpu_engine);
+}
 
-  auto in_mem          = in_data.GetMKLDNNData();
-  auto out_mem         = out_data.GetMKLDNNData(fwd.pd.dst_desc());
+void MKLDNNSoftmaxFwd::Execute(const Tensors& tensors) const {
   MKLDNNStream* stream = MKLDNNStream::Get();
-  stream->RegisterPrimArgs(fwd.GetFwd(), {{MKLDNN_ARG_SRC, *in_mem}, {MKLDNN_ARG_DST, *out_mem}});
+
+  auto original_input_mem = tensors.data.GetMKLDNNData();
+  const auto out_mem            = tensors.out.GetMKLDNNData(softmax_pd->dst_desc());
+
+  mkldnn::memory* softmax_input_mem;
+  if (temperature_pd) {  // temperature parameter used
+    // check whether additional buffer is needed
```

Comment:
Would not this comment fit into one line? E. g. `check whether additional buffer is needed, when temperature parameter is being used`